quinn-proto/src/token_store.rs

```diff
@@ -0,0 +1,18 @@
+use bytes::Bytes;
+
+/// Responsible for storing address validation tokens received from servers and retrieving them for
+/// use in subsequent connections
+pub trait TokenStore: Send + Sync {
+    /// Potentially store a token for later one-time use
+    ///
+    /// Called when a NEW_TOKEN frame is received from the server.
+    fn insert(&self, server_name: &str, token: Bytes);
+
+    /// Try to find and take a token that was stored with the given server name
+    ///
+    /// The same token must never be returned from `take` twice, as doing so can be used to
+    /// de-anonymize a client's traffic.
+    ///
+    /// Called when trying to connect to a server. It is always ok for this to return `None`.
+    fn take(&self, server_name: &str) -> Option<Bytes>;
```

Comment:
Calling `connect` with a server name and a trait object whose method is immediately called on the server name you passed in yourself seems needlessly complex. Should the token to use be a new optional parameter to `connect` or similar instead?

Reply:
On one hand, I essentially agree with you. On the other hand, I'm doing it this way so as to mirror the architecture of [`rustls::client::ClientSessionStore`](https://docs.rs/rustls/latest/rustls/client/trait.ClientSessionStore.html). Both out of an appeal to precedent, and because TLS session resumption and QUIC token usage must work in tandem for token usage to be effective.

I mean, if you think Rustls would be open to me creating a PR that deleted `rustls::client::ClientSessionStore` and instead provided an API to thread this through more directly, I'd be open to working on such a thing.

Reply:
> appeal to precedent

rustls started with a lot of synchronous APIs, but has recently gone to considerable effort to become less opinionated in that respect, as in e.g. acceptors. I don't think we should take the lingering synchronous patterns as gospel.

> TLS session resumption and QUIC token usage must work in tandem for token usage to be effective

Why? AFAICT they solve completely independent problems: TLS session resumption saves handshake CPU time, and QUIC address validation tokens save a round-trip.

> if you think Rustls would be open to me creating a PR

I wouldn't be surprised, but I can't speak with authority there. @djc?

Reply:
> AFAICT they solve completely independent problems: TLS session resumption saves handshake CPU time, and QUIC address validation tokens save a round-trip

Ok, I think we should clear up whatever confusion is happening here. Here's what I'm saying:

I don't think I'm mistaken about this: A QUIC client can only attempt to send 0-RTT data if it remembers certain QUIC transport parameters and TLS cryptographic material and maybe other information from TLS from a previous time it connected to that domain, right? As Quinn's 0-rtt docs state, "For outgoing connections, the initial attempt to convert to a [Connection](https://docs.rs/quinn/0.11.5/quinn/struct.Connection.html) which sends 0-RTT data will proceed if the [crypto::ClientConfig](https://docs.rs/quinn/0.11.5/quinn/crypto/trait.ClientConfig.html) attempts to resume a previous TLS session."

With the default way for things to work for Quinn, the sequence of things happening in the code base is this:

- The default `quinn::crypto::ClientConfig` is a `quinn::crypto::rustls::QuicClientConfig`
- `quinn::crypto::rustls::QuicClientConfig` wraps around a `rustls::client::ClientConfig`
- `rustls::client::ClientConfig` will attempt to resume a previous TLS session if its `resumption: rustls::client::Resumption` field says it will
- `rustls::client::Resumption` wraps around a `rustls::client::ClientSessionStore`
- The default `rustls::client::ClientSessionStore` is a `rustls::client::ClientSessionMemoryCache`

I believe that the TLS part of the process for being able to transmit 0-RTT data is sometimes referred to as TLS session resumption, or perhaps sometimes in this context, "stateless" session resumption in particular. Both the relevant QUIC docs and Rustls docs and types make reference to the word "resumption".

Reply:
Ah, right, I forgot that address validation tokens are most relevant when issuing large 0-RTT requests. That isn't technically the *only* time when they have an impact, but I'm willing to accept that it's the overwhelmingly important case.

Reply:
> I mean, if you think Rustls would be open to me creating a PR that deleted `rustls::client::ClientSessionStore` and instead provided an API to thread this through more directly, I'd be open to working on such a thing.

One design goal there is to let resumption work, by default, without significant work on the side of downstream application authors.

However, I'd be open to thinking about an alternative constructors for `quic::ClientConnection` that pass-in values that would otherwise be looked up? (Though that would make the API a bit lopsided since the new sessions/hints would still end up in the store...)

Reply:
> One design goal there is to let resumption work, by default, without significant work on the side of downstream application authors.

Very on board with this in theory, but I suspect most downstreams aren't often reconnecting to the same server from the same process, so I'm not sure how useful this is in practice.

Reply:
An argument to consider:

- Having an in-memory cache such as `ClientSessionMemoryCache` / `TokenMemoryCache` is cheap. It uses little memory and CPU.
- It being the default has the potential to give users automatic performance gains. Sure, many applications won't be reconnecting from the same process. But probably some will. Like long-running crawlers, for example.
- It is probably not possible to create a persistent cache without it being too opinionated for us to feel good about setting as the default.
- As ctz noted:

  > However, I'd be open to thinking about an alternative constructors for quic::ClientConnection that pass-in values that would otherwise be looked up? (Though that would make the API a bit lopsided since the new sessions/hints would still end up in the store...)

  Allowing a `TokenStore` / `ClientSessionStore` object is not mutually exclusive with allowing tokens / session resumption to be used "manually." For both Rustls and Quinn, we could implement:

  1. The ability to pass in a token / `ServerData` to the functions that initialize a client, such that Quinn / Rustls doesn't do a lookup on the `TokenStore` / `ClientSessionStore` if one is passed in manually.
  2. (Not technically strictly necessary): The ability to set the `TokenStore` / `ClientSessionStore` to `None`

  And then the user could still have the in-memory stores as their default, while also being able to take manual control without any superfluous runtime costs or significant API awkwardness.

This is an argument for allowing Rustls and Quinn's store objects with in-mem defaults to remain / proceed with implementation, and simply also add additional APIs in the future.

Reply:
> This is an argument for allowing Rustls and Quinn's store objects with in-mem defaults to remain / proceed with implementation, and simply also add additional APIs in the future.

Sounds good to me!

Reply:
I'll also note that a similar argument applies to possible concerns about whether we want the token log to be able to be async. Although as discussed in DMs I think, there are some more complications to doing that, what with path changes and associated race conditions. But it's still possible, and something I'm potentially interested in doing.

Reply:
> Very on board with this in theory, but I suspect most downstreams aren't often reconnecting to the same server from the same process, so I'm not sure how useful this is in practice.

You're probably right with "most", but there are plenty that do. As a random example: the rustls benchmarking bot is a long-lived process, and hits the github API a ton. (I don't have any data on how well TLS resumption improves the performance of that, and I'd expect HTTP connection pooling to also be in play and more effective for a burst of sequential requests.)

Reply:
I think it could be very common for processes to connect to the same server many times. For example, we had a large number of little HTTP API clients in our HTTP server processes that would maintain a pool of connections over time.